File: bitvm/vm.js
import { blake3 } from '../libs/blake3.js'
import { toHex, fromHex, concat } from '../libs/bytes.js'
import { buildTree, buildPath, verifyPath } from '../libs/merkle.js'
import {
    ASM_ADD,
    ASM_SUB,
    ASM_MUL,
    ASM_AND,
    ASM_OR,
    ASM_XOR,
    ASM_ADDI,
    ASM_SUBI,
    ASM_ANDI,
    ASM_ORI,
    ASM_XORI,
    ASM_JMP,
    ASM_BEQ,
    ASM_BNE,
    ASM_RSHIFT1,
    ASM_SLTU,
    ASM_SLT,
    ASM_LOAD,
    ASM_STORE,
    ASM_SYSCALL,
    PATH_LEN,
    TRACE_LEN,
    MEMORY_LEN,
    U32_SIZE
} from './constants.js'


// Map positive and negative n to an unsigned u32
export const toU32 = n => (U32_SIZE + (n % U32_SIZE)) % U32_SIZE

// A program is a list of instructions
export class Instruction {

    constructor(type, addressA, addressB, addressC) {
        this.type = type
        this.addressA = addressA
        this.addressB = addressB
        this.addressC = addressC
    }

    toString() {
        return `${this.type} ${this.addressA} ${this.addressB} ${this.addressC}`
    }
}

export const compileProgram = source => source.map(instruction => new Instruction(...instruction))


class MerklePath {
    #path
    #snapshot
    #address

    constructor(snapshot, address) {
        const memory = snapshot.memory
        if (address < 0)
            throw `ERROR: address=${address} is negative`
        if (address >= MEMORY_LEN)
            throw `ERROR: address=${address} >= MEMORY_LEN=${MEMORY_LEN}`
        this.#snapshot = snapshot
        this.#address = address
        this.#path = buildPath(memory.map(value => new Uint32Array([value]).buffer), address)
    }

    verifyUpTo(height) {
        height = PATH_LEN - height
        const subPath = this.#path.slice(0, height)
        const value = new Uint32Array([this.#snapshot.read(this.#address)]).buffer
        const node = verifyPath(subPath, value, this.#address)
        return toHex(node)
    }

    getNode(index) {
        index = PATH_LEN - 1 - index
        return toHex(this.#path[index])
    }
}


class Snapshot {
    pc
    memory
    stepCount = 0
    instruction

    constructor(memory, instruction, pc = 0) {
        this.memory = memory
        this.instruction = instruction
        this.pc = pc
    }

    read(address) {
        if (address < 0)
            throw `ERROR: address=${address} is negative`
        if (address >= MEMORY_LEN)
            throw `ERROR: address=${address} >= MEMORY_LEN=${MEMORY_LEN}`
        if (address >= this.memory.length)
            return 0
        return this.memory[address]
    }

    write(address, value) {
        if (address < 0)
            throw `ERROR: address=${address} is negative`
        if (address >= MEMORY_LEN)
            throw `ERROR: address=${address} >= MEMORY_LEN=${MEMORY_LEN}`
        this.memory[address] = value
    }
    
    path(address) {
        return new MerklePath(this, address)
    }

    get root() {
        const root = buildTree(this.memory.map(x => new Uint32Array([x]).buffer))
        return toHex(root)
    }
}

const executeInstruction = (s) => {

    //console.log(`PC: ${s.pc},  Instruction: ${(s.instruction+'').padEnd(9,' ')}, valueA: ${s.read(s.instruction.addressA)}, valueB: ${s.read(s.instruction.addressB)}, valueC: ${s.read(s.instruction.addressC)}`)
    switch (s.instruction.type) {
        case ASM_ADD:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) + s.read(s.instruction.addressB))
            )
            s.pc += 1
            break
        case ASM_SUB:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) - s.read(s.instruction.addressB))
            )
            s.pc += 1
            break
        case ASM_MUL:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) * s.read(s.instruction.addressB))
            )
            s.pc += 1
            break
        case ASM_AND:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) & s.read(s.instruction.addressB))
            )
            s.pc += 1
            break
        case ASM_OR:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) | s.read(s.instruction.addressB))
            )
            s.pc += 1
            break
        case ASM_XOR:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) ^ s.read(s.instruction.addressB))
            )
            s.pc += 1
            break
        case ASM_ADDI:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) + s.instruction.addressB)
            )
            s.pc += 1
            break
        case ASM_SUBI:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) - s.instruction.addressB)
            )
            s.pc += 1
            break
        case ASM_ANDI:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) & s.instruction.addressB)
            )
            s.pc += 1
            break
        case ASM_ORI:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) | s.instruction.addressB)
            )
            s.pc += 1
            break
        case ASM_XORI:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) ^ s.instruction.addressB)
            )
            s.pc += 1
            break
        case ASM_BEQ:
            if (s.read(s.instruction.addressA) == s.read(s.instruction.addressB)) {
                s.pc = s.instruction.addressC
            } else {
                s.pc += 1
            }
            break
        case ASM_BNE:
            if (s.read(s.instruction.addressA) != s.read(s.instruction.addressB)) {
                s.pc = s.instruction.addressC
            } else {
                s.pc += 1
            }
            break
        case ASM_JMP:
            s.pc = s.read(s.instruction.addressA)
            break
        case ASM_RSHIFT1:
            s.write(
                s.instruction.addressC,
                toU32(s.read(s.instruction.addressA) >>> 1)
            )
            s.pc += 1
            break
        case ASM_SLTU:
            s.write(s.instruction.addressC, s.read(s.instruction.addressA) >>> 0 < s.read(s.instruction.addressB) >>> 0 ? 1 : 0);
            s.pc += 1
            break            
        case ASM_SLT:
            // Binary OR with each value to cast them to 32-bit integer and then back to a sign-extended number.
            s.write(s.instruction.addressC, (s.read(s.instruction.addressA) | 0 ) < ( s.read(s.instruction.addressB) | 0 ) ? 1 : 0);
            s.pc += 1
            break
        case ASM_LOAD:
            s.instruction.addressA = s.read(s.instruction.addressB)
            s.write(s.instruction.addressC, s.read(s.instruction.addressA))
            s.pc += 1
            break
        case ASM_STORE:
            s.instruction.addressC = s.read(s.instruction.addressB)
            s.write(s.instruction.addressC, s.read(s.instruction.addressA))
            s.pc += 1
            break;
        case ASM_SYSCALL:
            console.log("syscall called")
            s.pc += 1
            break
        default:
            throw `Unsupported instruction type ${s.instruction.type}`
            // s.pc += 1
            // break
    }
}

export class VM {
    program
    memoryEntries

    constructor(programSource, memoryEntries) {
        this.program = compileProgram(programSource)
        this.memoryEntries = memoryEntries
    }

    run(maxSteps = TRACE_LEN) {
        const snapshot = new Snapshot([...this.memoryEntries], this.program[0])
        while (snapshot.pc < this.program.length && snapshot.stepCount < maxSteps) {
            snapshot.instruction = this.program[snapshot.pc]
            executeInstruction(snapshot)
            snapshot.stepCount++
        }
        return snapshot
    }

}

File: bitvm/src/vm.rs
use core::panic;

use crate::merkle::tree::{build_path, build_tree, verify_path};

use crate::constants::{
    ASM_ADD, ASM_ADDI, ASM_AND, ASM_ANDI, ASM_BEQ, ASM_BNE, ASM_JMP, ASM_LOAD, ASM_MUL, ASM_OR,
    ASM_ORI, ASM_RSHIFT1, ASM_SLT, ASM_SLTU, ASM_STORE, ASM_SUB, ASM_SUBI, ASM_SYSCALL, ASM_XOR,
    ASM_XORI, PATH_LEN,
};

#[derive(Copy, Clone)]
pub struct Instruction {
    pub asm_type: u8,
    pub address_a: u32,
    pub address_b: u32,
    pub address_c: u32,
}

pub struct Snapshot {
    pub pc: u32,
    pub memory: Vec<u32>,
    pub step_count: usize,
    pub instruction: Instruction,
}

pub struct MerklePath {
    pub path: Vec<[u8; 20]>,
    pub value: u32,
    pub address: u32,
}

impl MerklePath {
    fn new(snapshot: &Snapshot, address: u32) -> Self {
        Self {
            path: build_path(&snapshot.memory, address),
            value: snapshot.read(address),
            address,
        }
    }

    pub fn verify_up_to(&self, height: usize) -> [u8; 20] {
        let mut sub_path = self.path.clone();
        sub_path.shrink_to(PATH_LEN - height);
        verify_path(sub_path, self.value, self.address)
    }

    pub fn get_node(&self, index: usize) -> [u8; 20] {
        self.path[PATH_LEN - 1 - index]
    }
}

impl Snapshot {
    fn new(memory: Vec<u32>, instruction: Instruction, pc: u32) -> Self {
        Self {
            pc,
            memory,
            step_count: 0,
            instruction,
        }
    }

    pub fn read(&self, address: u32) -> u32 {
        self.memory[address as usize]
    }

    fn write(&mut self, address: u32, value: u32) {
        while address >= self.memory.len() as u32 {
            self.memory.push(0);
        }
        self.memory[address as usize] = value;
    }

    pub fn path(&self, address: u32) -> MerklePath {
        MerklePath::new(self, address)
    }

    pub fn root(&self) -> [u8; 20] {
        build_tree(&self.memory)
    }
}

fn execute_instruction(s: &mut Snapshot) {
    match s.instruction.asm_type {
        ASM_ADD => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a)
                    .wrapping_add(s.read(s.instruction.address_b)),
            );
            s.pc += 1
        }
        ASM_SUB => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a)
                    .wrapping_sub(s.read(s.instruction.address_b)),
            );
            s.pc += 1
        }
        ASM_MUL => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a)
                    .wrapping_mul(s.read(s.instruction.address_b)),
            );
            s.pc += 1
        }
        ASM_AND => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a) & s.read(s.instruction.address_b),
            );
            s.pc += 1
        }
        ASM_OR => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a) | s.read(s.instruction.address_b),
            );
            s.pc += 1
        }
        ASM_XOR => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a) ^ s.read(s.instruction.address_b),
            );
            s.pc += 1
        }
        ASM_ADDI => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a)
                    .wrapping_add(s.instruction.address_b),
            );
            s.pc += 1
        }
        ASM_SUBI => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a)
                    .wrapping_sub(s.instruction.address_b),
            );
            s.pc += 1
        }
        ASM_ANDI => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a) & s.instruction.address_b,
            );
            s.pc += 1
        }
        ASM_ORI => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a) | s.instruction.address_b,
            );
            s.pc += 1
        }
        ASM_XORI => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a) ^ s.instruction.address_b,
            );
            s.pc += 1
        }
        ASM_BEQ => {
            if s.read(s.instruction.address_a) == s.read(s.instruction.address_b) {
                s.pc = s.instruction.address_c
            } else {
                s.pc += 1
            }
        }
        ASM_BNE => {
            if s.read(s.instruction.address_a) != s.read(s.instruction.address_b) {
                s.pc = s.instruction.address_c
            } else {
                s.pc += 1
            }
        }
        ASM_JMP => s.pc = s.read(s.instruction.address_a),
        ASM_RSHIFT1 => {
            s.write(
                s.instruction.address_c,
                s.read(s.instruction.address_a) >> 1,
            );
            s.pc += 1
        }
        ASM_SLTU => {
            s.write(
                s.instruction.address_c,
                if s.read(s.instruction.address_a) < s.read(s.instruction.address_b) {
                    1
                } else {
                    0
                },
            );
            s.pc += 1
        }
        ASM_SLT => {
            s.write(
                s.instruction.address_c,
                if (s.read(s.instruction.address_a) as i32)
                    < (s.read(s.instruction.address_b) as i32)
                {
                    1
                } else {
                    0
                },
            );
            s.pc += 1
        }
        ASM_LOAD => {
            s.instruction.address_a = s.read(s.instruction.address_b);
            s.write(s.instruction.address_c, s.read(s.instruction.address_a));
            s.pc += 1
        }
        ASM_STORE => {
            s.instruction.address_c = s.read(s.instruction.address_b);
            s.write(s.instruction.address_c, s.read(s.instruction.address_a));
            s.pc += 1
        }
        ASM_SYSCALL => {
            println!("syscall called");
            s.pc += 1
        }
        _ => panic!("Unknown instuction type {}", s.instruction.asm_type),
    }
}

pub struct VM {
    program: Vec<Instruction>,
    memory_entries: Vec<u32>,
}

impl VM {
    pub fn new(program_source: &[Instruction], memory_entries: &[u32]) -> Self {
        Self {
            program: program_source.into(),
            memory_entries: memory_entries.into(),
        }
    }

    pub fn run(&self, max_steps: usize) -> Snapshot {
        let mut snapshot: Snapshot = Snapshot::new(self.memory_entries.clone(), self.program[0], 0);
        while snapshot.pc < self.program.len() as u32 && snapshot.step_count + 1 < max_steps {
            snapshot.instruction = self.program[snapshot.pc as usize];
            execute_instruction(&mut snapshot);
            snapshot.step_count += 1;
        }
        snapshot
    }
}
